File: app/src/main/java/quvesoft/project2/utils/RetrofitBuilder.java
package quvesoft.project2.utils;

import android.content.Context;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public final class RetrofitBuilder {
    public static <T> T create(Context context, final Class<T> service, final boolean loginNeeds, String token) {
        if (loginNeeds) {
            OkHttpClient.Builder httpClient = new OkHttpClient.Builder();

            httpClient.addInterceptor(chain -> {
                Request original = chain.request();

                Request request = original.newBuilder()
                        .header("Auth-Token", token)
                        .method(original.method(), original.body())
                        .build();

                return chain.proceed(request);
            });

            Retrofit retrofit = new Retrofit.Builder()
                    .baseUrl("http://52.78.223.19/chef/")
                    .addConverterFactory(GsonConverterFactory.create())
                    .client(httpClient.build())
                    .build();

            return retrofit.create(service);
        } else {
            Retrofit retrofit = new Retrofit.Builder()
                    .baseUrl("http://52.78.223.19/chef/")
                    .addConverterFactory(GsonConverterFactory.create())
                    .build();

            return retrofit.create(service);
        }
    }

    public static <T> T create(Context context, final Class<T> service, boolean loginNeeds) {
        return create(context, service, loginNeeds, "");
    }
}

File: app/src/main/java/com/yhjoo/dochef/data/network/RetrofitBuilder.kt
package com.yhjoo.dochef.data.network

import android.content.Context
import com.yhjoo.dochef.R
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.util.concurrent.TimeUnit

object RetrofitBuilder {
    private val retrofitClient: OkHttpClient by lazy {
        OkHttpClient.Builder()
            .connectTimeout(3, TimeUnit.SECONDS)
            .writeTimeout(10, TimeUnit.SECONDS)
            .readTimeout(10, TimeUnit.SECONDS)
            .build()
    }

    fun <T> create(context: Context, service: Class<T>): T {
        return Retrofit.Builder()
            .client(retrofitClient)
            .baseUrl(context.getString(R.string.server_url))
            .addConverterFactory(GsonConverterFactory.create())
            .build()
            .create(service)
    }
}